packages/gas-fee-controller/src/GasFeeControllerPolling.ts

```diff
@@ -0,0 +1,97 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseControllerV2 } from '@metamask/base-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+import type { NetworkClientId } from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+import { v1 as random } from 'uuid';
+
+export type PollingCompleteType<N extends string> = {
+  type: `${N}:pollingComplete`;
+  payload: [string];
+};
+
+export default abstract class GasFeeControllerPolling<
+  N extends string,
+  S extends Record<string, Json>,
+  messenger extends RestrictedControllerMessenger<
+    N,
+    any,
+    PollingCompleteType<N> | any,
+    string,
+    string
+  >,
+> extends BaseControllerV2<N, S, messenger> {
+  private readonly intervalLength = 1000;
+
+  private readonly networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
+    new Map();
+
+  private readonly intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
+
+  start(networkClientId: NetworkClientId) {
+    const innerPollToken = random();
+    if (this.networkClientIdTokensMap.has(networkClientId)) {
+      //
+      const set = this.networkClientIdTokensMap.get(networkClientId);
+      set?.add(innerPollToken);
+    } else {
+      const set = new Set<string>();
+      set.add(innerPollToken);
+      this.networkClientIdTokensMap.set(networkClientId, set);
+    }
+    this.#poll(networkClientId);
+
+    // call _poll
+    // add the inner poll token to the poll tokens set
+    return innerPollToken;
+  }
+
+  stop({
```

Comment:
Should we split this out into a stop & a seperate `stopAll` method, and tighten up this interface?

Also, should we be able to stop all pollling for a given network? (perhaps by omitting the polling token and only giving the networkCLientId?)

Also, should we be able to stop polling for a token, and not require passing in the networkCLientId?